File: staylio/src/services/hotelService.js
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_URL;

const hotelApi = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add request interceptor for auth token if needed
hotelApi.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Hotel Service
const hotelService = {
  // Get all hotels
  getAllHotels: async () => {
    const response = await hotelApi.get('/hotels');
    return response.data;
  },

  // Get landing page hotels
  getLandingPageHotels: async () => {
    const response = await hotelApi.get('/hotels/landing');
    return response.data;
  },

  // Get hotel by ID
  getHotelById: async (id) => {
    const response = await hotelApi.get(`/hotels/${id}`);
    return response.data;
  },

  // Get active hotels
  getActiveHotels: async () => {
    const response = await hotelApi.get('/hotels/active');
    return response.data;
  },

  // Get featured hotels
  getFeaturedHotels: async () => {
    const response = await hotelApi.get('/hotels/featured');
    return response.data;
  },

  // Search hotels
  searchHotels: async (keyword) => {
    const response = await hotelApi.get('/hotels/search', {
      params: { keyword },
    });
    return response.data;
  },

  // Get hotels by city
  getHotelsByCity: async (city) => {
    const response = await hotelApi.get(`/hotels/city/${city}`);
    return response.data;
  },

  // Get hotels by state
  getHotelsByState: async (state) => {
    const response = await hotelApi.get(`/hotels/state/${state}`);
    return response.data;
  },

  // Get hotels by country
  getHotelsByCountry: async (country) => {
    const response = await hotelApi.get(`/hotels/country/${country}`);
    return response.data;
  },

  // Get hotels by price range
  getHotelsByPriceRange: async (minPrice, maxPrice) => {
    const response = await hotelApi.get('/hotels/price-range', {
      params: { minPrice, maxPrice },
    });
    return response.data;
  },

  // Get hotels by minimum rating
  getHotelsByRating: async (minRating) => {
    const response = await hotelApi.get(`/hotels/rating/${minRating}`);
    return response.data;
  },

  // Get hotels by guest capacity
  getHotelsByGuestCapacity: async (guests) => {
    const response = await hotelApi.get(`/hotels/capacity/${guests}`);
    return response.data;
  },

  // Get hotels with available rooms
  getAvailableHotels: async () => {
    const response = await hotelApi.get('/hotels/available');
    return response.data;
  },

  // Get hotel images by ID
  getHotelImages: async (hotelId) => {
    const response = await hotelApi.get(`/hotel_img/${hotelId}`);
    return response.data;
  },

  // Get nearby hotels
  getNearbyHotels: async (latitude, longitude, radius = 10) => {
    const response = await hotelApi.get('/hotels/nearby', {
      params: { latitude, longitude, radius },
    });
    return response.data;
  },

  // Create hotel (admin/host only)
  createHotel: async (hotelData) => {
    const response = await hotelApi.post('/hotels', hotelData);
    return response.data;
  },

  // Update hotel (admin/host only)
  updateHotel: async (id, hotelData) => {
    const response = await hotelApi.put(`/hotels/${id}`, hotelData);
    return response.data;
  },

  // Delete hotel (admin/host only)
  deleteHotel: async (id) => {
    const response = await hotelApi.delete(`/hotels/${id}`);
    return response.data;
  },

  // Check if hotel is claimed
  checkHotelClaimStatus: async (hotelId) => {
    const response = await hotelApi.get(`/hotels/${hotelId}/claim-status`);
    return response.data;
  },

  // Get hotel reviews
  getHotelReviews: async (hotelId) => {
    const response = await hotelApi.get(`/reviews/hotel/${hotelId}`);
    return response.data;
  },

  // Add hotel review
  addHotelReview: async (reviewData) => {
    // Uses new validation endpoint
    const response = await hotelApi.post(`/hotels/${reviewData.hotelId}/reviews`, reviewData);
    return response.data;
  },

  // Check review eligibility
  checkReviewEligibility: async (hotelId, userId) => {
    const response = await hotelApi.get(`/hotels/${hotelId}/review-eligibility`, {
      params: { userId }
    });
    return response.data;
  },

  // Get hotel rooms
  getHotelRooms: async (hotelId) => {
    const response = await hotelApi.get(`/hotels/${hotelId}/hotel-rooms/available`);
    return response.data;
  },
};

export default hotelService;

File: staylio-backend/src/main/java/com/staylio/backend/Service/HotelService.java
package com.staylio.backend.Service;

import com.staylio.backend.Repo.HotelRepository;
import com.staylio.backend.model.Hotel;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class HotelService {

    @Autowired
    private HotelRepository hotelRepository;

    // Create hotel
    public Hotel createHotel(Hotel hotel) {
        if (hotel.getId() == null) {
            hotel.setId(generateUniqueId());
        }
        return hotelRepository.save(hotel);
    }

    private Long generateUniqueId() {
        long id;
        do {
            // Generate a random 6-digit number (100000 to 999999)
            id = 100000L + (long) (Math.random() * 900000L);
        } while (hotelRepository.existsById(id));
        return id;
    }

    // Get all hotels
    public List<Hotel> getAllHotels() {
        return hotelRepository.findAllSortedByClaimStatus();
    }

    // Get landing page hotels
    public List<Hotel> getLandingPageHotels() {
        return hotelRepository.findTop15ByHotelOwnerIdIsNotNullAndIsActiveTrueOrderByRatingDesc();
    }

    // Get hotel by ID
    public Optional<Hotel> getHotelById(Long id) {
        return hotelRepository.findById(id);
    }

    // Update hotel
    public Hotel updateHotel(Long id, Hotel hotelDetails) {
        Hotel hotel = hotelRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Hotel not found with id: " + id));

        hotel.setName(hotelDetails.getName());
        hotel.setDescription(hotelDetails.getDescription());
        hotel.setAddress(hotelDetails.getAddress());
        hotel.setCity(hotelDetails.getCity());
        hotel.setState(hotelDetails.getState());
        hotel.setCountry(hotelDetails.getCountry());
        hotel.setZipCode(hotelDetails.getZipCode());
        hotel.setLatitude(hotelDetails.getLatitude());
        hotel.setLongitude(hotelDetails.getLongitude());
        hotel.setPricePerNight(hotelDetails.getPricePerNight());
        hotel.setTotalRooms(hotelDetails.getTotalRooms());
        hotel.setAvailableRooms(hotelDetails.getAvailableRooms());
        hotel.setBedrooms(hotelDetails.getBedrooms());
        hotel.setBathrooms(hotelDetails.getBathrooms());
        hotel.setMaxGuests(hotelDetails.getMaxGuests());
        hotel.setPropertyType(hotelDetails.getPropertyType());
        hotel.setRoomTypes(hotelDetails.getRoomTypes());
        hotel.setAmenities(hotelDetails.getAmenities());
        hotel.setCheckInTime(hotelDetails.getCheckInTime());
        hotel.setCheckOutTime(hotelDetails.getCheckOutTime());
        hotel.setCancellationPolicy(hotelDetails.getCancellationPolicy());
        hotel.setHouseRules(hotelDetails.getHouseRules());
        hotel.setIsActive(hotelDetails.getIsActive());
        hotel.setIsFeatured(hotelDetails.getIsFeatured());

        return hotelRepository.save(hotel);
    }

    // Delete hotel
    public void deleteHotel(Long id) {
        Hotel hotel = hotelRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Hotel not found with id: " + id));
        hotelRepository.delete(hotel);
    }

    // Get hotels by host
    public List<Hotel> getHotelsByHost(Long hostId) {
        return hotelRepository.findByHostId(hostId);
    }

    // Get hotels by owner
    public List<Hotel> getHotelsByOwner(Long ownerId) {
        return hotelRepository.findByHotelOwnerId(ownerId);
    }

    // Get active hotels
    public List<Hotel> getActiveHotels() {
        return hotelRepository.findByIsActiveTrue();
    }

    // Get featured hotels
    public List<Hotel> getFeaturedHotels() {
        return hotelRepository.findByIsFeaturedTrue();
    }

    // Get hotels by city
    public List<Hotel> getHotelsByCity(String city) {
        return hotelRepository.findByCityIgnoreCaseAndIsActiveTrue(city);
    }

    // Get hotels by state
    public List<Hotel> getHotelsByState(String state) {
        return hotelRepository.findByStateIgnoreCase(state);
    }

    // Get hotels by country
    public List<Hotel> getHotelsByCountry(String country) {
        return hotelRepository.findByCountryIgnoreCase(country);
    }

    // Get hotels by price range
    public List<Hotel> getHotelsByPriceRange(BigDecimal minPrice, BigDecimal maxPrice) {
        return hotelRepository.findByPriceRange(minPrice, maxPrice);
    }

    // Get hotels by minimum rating
    public List<Hotel> getHotelsByMinimumRating(Double minRating) {
        return hotelRepository.findByMinimumRating(minRating);
    }

    // Get hotels by guest capacity
    public List<Hotel> getHotelsByGuestCapacity(Integer guests) {
        return hotelRepository.findByGuestCapacity(guests);
    }

    // Search hotels
    public List<Hotel> searchHotels(String keyword) {
        return hotelRepository.searchHotels(keyword);
    }

    // Get hotels with available rooms
    public List<Hotel> getHotelsWithAvailableRooms() {
        return hotelRepository.findHotelsWithAvailableRooms();
    }

    // Get nearby hotels
    public List<Hotel> getNearbyHotels(Double latitude, Double longitude, Double radius) {
        return hotelRepository.findNearbyHotels(latitude, longitude, radius);
    }

    // Update hotel rating
    public Hotel updateHotelRating(Long id, Double rating, Integer reviewCount) {
        Hotel hotel = hotelRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Hotel not found with id: " + id));
        hotel.setRating(rating);
        hotel.setReviewCount(reviewCount);
        return hotelRepository.save(hotel);
    }
}
